packages/fxa-metrics-processor/src/test/lib/pubsub.spec.ts

```diff
@@ -0,0 +1,205 @@
+        queueProcessor.pubSubClient = {
+          topic: sandbox.stub().returns({ publishJSON: publishStub })
+        };
+        await queueProcessor.processBatch();
+        sinon.assert.calledOnce(queueProcessor.pubSubClient.topic);
+        sinon.assert.calledOnce(queueProcessor.subClient.acknowledge);
+        sinon.assert.calledOnce(publishStub);
+        sinon.assert.calledWithExactly(queueProcessor.subClient.acknowledge, {
+          subscription: 'projects/default-project/subscriptions/fxa-metrics-raw-events',
+          ackIds: [1234]
+        });
+        sinon.assert.calledWithExactly(publishStub, { message: 'test' });
+      });
+
+      it('should not publish filtered messages', async () => {
+        queueProcessor.messageProcessor = { processMessage: sandbox.stub().returns(null) };
+        await queueProcessor.processBatch();
+        sinon.assert.notCalled(queueProcessor.pubSubClient.topic);
+        sinon.assert.calledOnce(queueProcessor.subClient.acknowledge);
+        sinon.assert.calledWithExactly(queueProcessor.subClient.acknowledge, {
+          subscription: 'projects/default-project/subscriptions/fxa-metrics-raw-events',
+          ackIds: [1234]
+        });
+      });
+
+      it('should handle errors processing messages', async () => {
+        queueProcessor.messageProcessor = { processMessage: sandbox.stub().throws() };
+        queueProcessor.handleMessageError = sinon.stub().resolves(true);
+        await queueProcessor.processBatch();
+        sinon.assert.notCalled(queueProcessor.pubSubClient.topic);
+        sinon.assert.calledOnce(queueProcessor.subClient.acknowledge);
+        sinon.assert.calledWithExactly(queueProcessor.subClient.acknowledge, {
+          subscription: 'projects/default-project/subscriptions/fxa-metrics-raw-events',
+          ackIds: [1234]
+        });
+      });
+
+      it('should handle errors that fail to publish to dead letter queue', async () => {
+        queueProcessor.messageProcessor = { processMessage: sandbox.stub().throws() };
+        queueProcessor.handleMessageError = sinon.stub().resolves(false);
+        await queueProcessor.processBatch();
+        sinon.assert.notCalled(queueProcessor.pubSubClient.topic);
+        sinon.assert.calledOnce(queueProcessor.subClient.acknowledge);
+        sinon.assert.calledWithExactly(queueProcessor.subClient.acknowledge, {
+          subscription: 'projects/default-project/subscriptions/fxa-metrics-raw-events',
+          ackIds: []
+        });
+      });
+    });
+
+    describe('start and stop', () => {
+      it('works', async () => {
+        const emitter = new EventEmitter();
+        queueProcessor.processBatch = () => once(emitter, 'next');
+        assert.equal(queueProcessor.shouldStop, false);
+        // Don't wait for the promise
+        queueProcessor.start();
+
+        // Iterate the loop and yield event loop so it runs
+        emitter.emit('next');
+        await sleep(0);
```

Comment:
Setting the timeout like this is fine, though sinon does provide a [mock clock](https://sinonjs.org/releases/v1.17.6/fake-timers/) that can be used to advance time in a more controlled way.

Reply:
This isn't being used to timeout, it's being used to yield the event loop so the listener waiting on the emitter event can be run. The mock clock does not yield the event loop, its synchronous.